specs/discussions/2025-04-block-proofs.md

```diff
@@ -0,0 +1,319 @@
+## Shard blocks commitment and proof generation.
+
+The goal of this PR is to surface the core data structures and mechanics of the process of
+committing and verifying the history of a child shard to the upper levels of the hierarchy.
+
+_NOTE: This PR is for discussion purposes only and I don't expect it to be merged as is. Eventually
+this will become part of the implementation and/or the final spec._
+
+### Submitting blocks to the parent
+
+1. Assume the following information (at least) in the `BlockHeader` of a shard block. The fields
+   included from the canonical block structure from any other blockchain are:
+   - A reference to the latest final block seen by the farmer in the beacon chain when proposing
+     this block.
+   - The root hash of all the blocks and segments from the child shard being submitted to the upper
+     layers in this block (more about this in the sections below).
+   - The root hash of the state of the child shard after applying the transactions in this block.
+   - Raw consensus information that needs to be included in a block to submitted to the upper layers
+     (e.g. `SegmentDescription`)
+   - The root hash of the Mountain Merkle Tree (MMR) of the history of the shard after appending the
+     parent block.
+
+```rust
+struct BlockHeader<T: SegmentDescription>
+	/// Block number
+	number: BlockNumber
+	/// State root (can be considered redundant if tx_root is included)
+	state_root: Hash
+	/// Root of the Merkle tree of transactions included in the block.
+	tx_root: Hash
+	/// Block solution.
+	solution: Proof
+	/// Block number and hash of the beacon block referenced by this hash block.
+	beacon_chain_ref: (BlockNumber, Hash)
+	// The commitment of consensus information for a shard is not stored as
+	// plain transactions in the block, but as their raw data structures (mainly block
+	// and segment information from the child shard).
+	consensus_info_root: Hash
+	/// List of segments and consensus objects that need to be submitted to the upper
+	/// layers of the chain.
+	consensus_info: Vec<T>
```

Comment:
Why is this generic, isn't it supposed to be a fixed known data structure?

Reply:
I see. At least in Rust generics always represent a single type. It can be one or the other, but not both. So if you want to store it in a vector, it'll have to either be `Box<dyn T>` or you create an enum with different variants representing different things.

That said, I think Substrate's notion of digest that I mentioned in a different thread fits nicely here.

Reply:
You are right, `Digest` works like a charm. Thanks for suggesting it. Changed it.